File: sections/61_icpc_contest_readiness/problems/b_time_boxed_schedule/solution.py
import heapq
import sys


def main() -> None:
    data = list(map(int, sys.stdin.buffer.read().split()))
    if not data:
        return
    n = data[0]
    tasks = [(data[i], data[i + 1]) for i in range(1, 2 * n + 1, 2)]
    tasks.sort(key=lambda task: task[1])
    chosen: list[int] = []
    total = 0
    for duration, deadline in tasks:
        total += duration
        heapq.heappush(chosen, -duration)
        if total > deadline:
            total += heapq.heappop(chosen)
    print(len(chosen))


if __name__ == "__main__":
    main()

File: sections/61_icpc_contest_readiness/problems/b_time_boxed_schedule/solution.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n;
    if (!(cin >> n)) return 0;
    vector<pair<long long, long long>> tasks(n);
    for (auto& [duration, deadline] : tasks) cin >> duration >> deadline;
    sort(tasks.begin(), tasks.end(), [](auto a, auto b) { return a.second < b.second; });

    priority_queue<long long> chosen;
    long long total = 0;
    for (auto [duration, deadline] : tasks) {
        total += duration;
        chosen.push(duration);
        if (total > deadline) {
            total -= chosen.top();
            chosen.pop();
        }
    }
    cout << chosen.size() << '\n';
}
